go/mathutil/equivalence_relation.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mathutil
+
+import (
+	"fmt"
+	"sort"
+)
+
+type UnknownElementError struct {
+	element string
+}
+
+func (e *UnknownElementError) Error() string {
+	return fmt.Sprintf("unknown element %s", e.element)
+}
+
+type UnknownClassError struct {
+	class int
+}
+
+func (e *UnknownClassError) Error() string {
+	return fmt.Sprintf("unknown class %d", e.class)
+}
+
+// EquivalenceRelation implements a mathematical equivalence relation.
+// Elements in this set are named, ie identified by strings.
+// Elements are potentially grouped together in an equivalence relation. Each element belongs in exactly one
+// relation, and each relation has 1 or more elements.
+// If a,b are in the same relation, and if b,c are in the same relation, it follows that a,c are in the same relation.
+// therefore two different entity relations cannot have any shared elements.
+// Functions of this struct are not thread safe.
+type EquivalenceRelation struct {
+	elementClassMap  map[string]int
+	classElementsMap map[int]([]string)
+
+	classCounter int
+}
+
+func NewEquivalenceRelation() *EquivalenceRelation {
+	return &EquivalenceRelation{
+		elementClassMap:  make(map[string]int),
+		classElementsMap: make(map[int][]string),
+	}
+}
+
+// Add adds a single element to the set. The element is associated with its own unique class
+func (r *EquivalenceRelation) Add(element string) {
+	if _, ok := r.elementClassMap[element]; ok {
+		// element already exists
+		return
+	}
+	r.elementClassMap[element] = r.classCounter
+	r.classElementsMap[r.classCounter] = []string{element}
+	r.classCounter++
```

Comment:
Right. As the equivalence relation goes, it's only important that any new element gets its own distinct class-number. So `r.classCounter++` ensures the next element is added onto its own distinct class.

The users of the equivalence relation, though, are interested in the order, just so that we have kind of a stable sorting mechanism.